http/events.go

```diff
@@ -0,0 +1,129 @@
+package http
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+	"google.golang.org/protobuf/encoding/protojson"
+	"nhooyr.io/websocket"
+)
+
+// handleEventsSubscribeWebsocket runs forever, returning a websocket error code and reason
+// only if the connection closes or there was an error.
+func handleEventsSubscribeWebsocket(ctx context.Context, core *vault.Core, ns *namespace.Namespace, eventType logical.EventType, conn *websocket.Conn, json bool) (websocket.StatusCode, string, error) {
+	events := core.Events()
+	ch, cancel, err := events.Subscribe(ctx, ns, eventType)
+	if err != nil {
+		core.Logger().Info("Error subscribing", "error", err)
+		return websocket.StatusUnsupportedData, "Error subscribing", nil
+	}
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			core.Logger().Info("Websocket context is done, closing the connection")
+			return websocket.StatusNormalClosure, "", nil
+		case message := <-ch:
+			core.Logger().Debug("Sending message to websocket", "message", message)
+			var messageBytes []byte
+			if json {
+				messageBytes, err = protojson.Marshal(message)
+			} else {
+				messageBytes, err = proto.Marshal(message)
+			}
+			if err != nil {
+				core.Logger().Warn("Could not serialize websocket event", "error", err)
+				return 0, "", err
+			}
+			messageString := string(messageBytes) + "\n"
+			err = conn.Write(ctx, websocket.MessageText, []byte(messageString))
+			if err != nil {
+				return 0, "", err
+			}
+		}
+	}
+}
+
+func handleEventsSubscribe(core *vault.Core) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		core.Logger().Debug("Got request to", "url", r.URL, "version", r.Proto)
+
+		ctx := r.Context()
+		ns, err := namespace.FromContext(ctx)
+		if err != nil {
+			core.Logger().Info("Could not find namespace", "error", err)
+			respondError(w, http.StatusInternalServerError, fmt.Errorf("could not find namespace"))
+			return
+		}
+
+		prefix := "/v1/sys/events/subscribe/"
+		if ns.ID != "root" {
+			prefix = fmt.Sprintf("/v1/%s/sys/events/subscribe/", ns.Path)
+		}
+		eventTypeStr := strings.TrimSpace(strings.TrimPrefix(r.URL.Path, prefix))
+		if eventTypeStr == "" {
```

Comment:
Thinking of other possible error cases, I don't think it should be possible that we didn't trim any prefix at all right? Otherwise we shouldn't have got into this handler.

Reply:
I believe that is correct. I did some quick manual testing and I wasn't able to get this to misbehave.